Canton Party Resolution Protocol - PR (#171)
global-synchronizer-foundation/cips: cprp-resolution.md

Number: CIP-XXXX

Title: Canton Party Name Resolution Standard

Status: Draft

Author: Paolo Domenighetti (Freename), [Co-authors TBD]

Created: 2026-08-27

## Abstract

> This CIP standardizes four things and deliberately nothing more: how human-readable party names are represented on the Canton Network (the Fully Qualified Party Name format), how they are resolved to Canton Party IDs, how they are rendered in application UIs, and how additional name sources are integrated by defining new resolvers or extending existing ones.

Reviewer: ```suggestion
This CIP defines a uniform approach for naming parties on the Canton network.
The approach allows integrating the many naming systems used by the organizations on the Canton Network regardless of whether they are native to the Canton Network (like CNS) or externally managed (like DNS or LEI).
The approach is based on standardizing the following four concepts as part of this CIP: how human-readable party names are represented on the Canton Network (the Fully Qualified Party Name format), how they are resolved to Canton Party IDs, how they are rendered in application UIs, and how additional naming systems are integrated by defining new resolvers or extending existing ones.
```

Note: let's decide on "naming system" over "name source" and use consistently. I'm leaning towards the former as it seems more general.

> It aligns with the CNS 2.0 architecture proposed by Simon Meier (Digital Asset) and the Working Group direction converged over the May–August 2026 meetings.

Reviewer: Push to rationale. Make the CIP earn its merit based on its text.

## Motivation

> Canton participants are identified by cryptographic Party IDs of the form `<prefix>::<68-character-namespace-hash>` — unusable in human workflows. CNS 1.0 names (`name.unverified.cns`) are self-registered without verification and convey no signal about which counterparty they identify. Applications need to display trustworthy human-readable names, accept them as input, and discover metadata keyed on a party — uniformly across the network. This CIP provides the shared representation, resolution behavior, and rendering conventions that make that possible, while leaving every implementation choice to implementers.

Reviewer: ```suggestion
Canton Network participants are identified by cryptographic Party IDs of the form `<prefix>::<68-character-namespace-hash>`, which are unusable in human workflows other than simple copy-and-pasting. CNS 1.0 names (`name.unverified.cns`) are self-registered without verification and convey no signal about which counterparty they identify. Applications need to display memorable human-readable names and accept them as input uniformly across the network. This CIP provides the shared representation, resolution behavior, and rendering conventions that make that possible, while leaving ample implementation freedom to implementers.
```

#### 2.1 Format

```
<network>:<resolver>:<name>
```

| Component | Purpose | Values |
|-----------|---------|--------|
| `network` | Prevents cross-environment confusion | `mainnet`, `testnet`, `devnet`, `localnet` |
| `resolver` | The naming system that backs this FQPN | `dns`, `lei`, `ens`, `cns`, `party-id`, others via their own CIPs |
| `name` | A sequence of printable, non-whitespace ASCII characters, interpreted by the resolver | `blackrock.com`, `blackrock.canton`, `partyHint::1220abcd...` |

Examples:

- `mainnet:dns:blackrock.com`
- `mainnet:dns:tmmf.lloyds.com`
- `mainnet:cns:blackrock.canton`
- `mainnet:lei:784F5XWPLTWKTBV3E584`
- `mainnet:party-id:partyHint::1220abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000`

The `<name>` component is opaque to this standard and interpreted only by the resolver. This lets any external naming syntax — DNS dots, LEI codes, Party IDs with their `::` separators — embed without escaping.

#### 2.2 Network Discrimination

The `network` component is mandatory. TestNet names MUST NOT be confusable with MainNet names. Applications supporting multiple networks SHOULD display the network.

> #### 2.3 The `party-id` Resolver (Built-in)

Reviewer: this is a resolution aspect not a representation aspect

The `party-id` resolver is defined by this CIP and available on every network. Its `<name>` is the Canton Party ID verbatim. It guarantees that every Canton party always has at least one FQPN, before any name registration, with no external state and no trust beyond the Party ID itself.

### 3. Resolution

#### 3.1 Resolver Operations

Every resolver provides two logical operations. How they are transported and exposed is implementation-defined.

- resolve: name → the Party ID the name maps to, plus associated metadata claims and a validity horizon.

> - reverseResolve: Party ID → the names this resolver knows for it.

Reviewer: Here I'm inclined to give an actual HTTP OpenAPI spec. Otherwise name providers and consumers cannot act independently and have to do more coordination. Removing this coordination need is the goal of these standardization CIPs.

#### 3.2 Forward Resolution

An application resolving a user-typed name matches it against each resolver in its configured, ordered list and returns the first successful result. Per-resolver ignore rules (patterns on the `<name>` component) MAY skip a resolver for matching names, addressing overlaps between resolvers when the application wants a specific delegation.

#### 3.3 Preferred-Name Resolution (Reverse Direction)

Given a Party ID, an application computes the preferred FQPN to display:

> 1. Retrieve the party's preferred-name credential from the Credentials Registry: issuer = holder = the party, claim `cip-cns/preferred-name` (claim key per the CNS 2.0 CIP).

Reviewer: Interesting. I wasn't aware of this dependency. Makes sense to have. Note though that it's not the CNS CIP, but this CIP that defines how parties publish their preferred names. 

That said: I'm kind of inclined to actually push this into the party profile CIP. It seems like another kind of self-published party metadata. wdyt?

@waynecollier-da wdyt?

Reviewer: Note: @pdomenighetti and myself agree that this should be pushed to party profiles.

2. Interpret the claim value as an FQPN and resolve it forward per Section 3.2, verifying it resolves back to the same Party ID. This round-trip verification prevents a party from claiming a name it does not hold.
3. If verified, display that name.
4. Otherwise, perform reverse resolution across the configured resolvers and select one result by the application's preference order — by default, lexicographic sort by (application priority override, user priority, application priority default).

> ### 4. Rendering

Reviewer: I'd suggest to move this right after the representation. I think it would also be good to have two subsections: one for the ASCII Rendering, which just states that it is verbatim, and one for the GUI Rendering.

Rendered names follow a shared convention: a source icon followed by the human-readable name.

```
<icon> <display-name>
```

- For `dns`, `lei`, `ens`, and `party-id`, there is one standardized icon per resolver.

> - For `cns`, the icon is the logo of the TLD under which the name is registered. Registrars self-publish their TLD logo, registered via a DSO-issued credential (`cip-cns/tld-logo-url`, per the CNS 2.0 CIP); applications retrieve the logo URL from the Credentials Registry at render time.

Reviewer: Avoid the forward reference. Instead structure the icon resolution algorithm such that resolvers can specify both how the icon and the name is determined.

Reviewer: Also: I'd suggest to add a screenshot of the Etherscan and its <icon> <name> format in the rationale to justify this decision giving prior art.

- Fallback chain when no name resolves: CNS 1.0 entry → Party ID prefix → truncated Party ID.

Assets are not named by FQPNs. An instrument identified by the token standard's `InstrumentId {admin: Party, id: Text}` renders as `<symbol> by <admin-rendered-name>` (e.g. `CC by <cns icon> dso.cns`), where the symbol is published by the admin through its token-standard metadata endpoint (`/registry/metadata/v1/instruments/{id}`), discovered via the admin's credentials (see the forthcoming CIP-56 amendment).

### 5. Integrating Additional Name Sources

New name sources enter the Canton Network either by defining a new resolver or by extending an existing one, in each case via a dedicated CIP. Such a CIP MUST specify:

- The resolver identifier as it appears in FQPNs (or the existing resolver being extended).
- The `<name>` syntax the resolver accepts.
- The naming authority and the governance under which its registrars operate.
- The procedure by which name-to-Party-ID bindings are established and, where the underlying authority can change externally, the re-verification cadence.

> - The on-ledger credential encoding of name registrations, using the CN Credentials Standard. Names MUST be materialized as on-ledger credentials so that Daml code can reference them (import model, per the July 9, 2026 WG direction); reference-only integration without on-ledger state is acceptable only where no Daml logic depends on the name.

Reviewer: I would make this optional. Names have ample value for display purposes only.

Reviewer: I'm even inclined to completely remove this CIP. It seems to be more of an aspect of machine-checked identity verification than an aspect of making it easy for humans to interact with party-ids.

- The rendered-name convention, if it differs from `<name>` verbatim.

Registration credentials SHOULD carry at minimum the canonical FQPN, the network, and a validity horizon as claims. The claim-key prefix used by this document family (`cprp/`) is a working prefix, to be renamed `cip-<nr>/` on number assignment, consistent with `cip-cns/*`.

> The first such source CIP is the Canton Imported Names CIP (Decentralized DNS Name Importer), which specifies DNS import in full and serves as the blueprint for LEI/vLEI, ENS, email, and further importers.

Reviewer: Don't precommit to another CIP in this CIP ;-)

Reviewer: let's remove this

- Implements the "apps decide resolution strategy" principle — no foundation-mandated resolution policy
- Avoids bloating the ACS of the DSO party (resolution queries are off-ledger)
- Avoids bloating the Scan API surface (additive changelog integration only)
- Supports integrating existing name providers (DNS, ENS, vLEI) as resolvers
- Supports local address books as a resolver type

## Specification Overview

The full specification is in the shared companion document [CPRP-spec.md](./CPRP-spec.md). This section summarizes the resolution-layer design elements.

### Fully Qualified Party Name (FQPN)

Names are structured as `<network>/<resolver>:<namespace>:<n>`:

| Component | Purpose | Example Values |
|-----------|---------|---------------|
| `network` | Prevents cross-environment confusion | `mainnet`, `testnet`, `devnet` |
| `resolver` | Identity source that backs the name | `dns`, `vlei`, `freename`, `self` |
| `namespace` | Registrant's domain or organizational scope | `goldmansachs.com`, `acme-bank.canton` |
| `n` | Specific name within the namespace | `default`, `treasury`, `trading-desk-3` |

Examples:

> - `mainnet/dns:goldmansachs.com:default`

Reviewer: consider avoiding the `mainnet/` prefix, as parties are not bound to networks. You might use `party:` without a qualifier or use `protocol/party`. Or perhaps avoid even stating that as party and just use the Canton concept: `uid:`, which stands for Universal Identifier, and is cryptic enough so that people don't attach too many unwarranted concepts to it.

- `mainnet/vlei:784F5XWPLTWKTBV3E584:default`
- `testnet/freename:acme-bank.canton:treasury`

Network discrimination is a security requirement: TestNet names must not be confusable with MainNet names.

### Resolver Interface

Any identity provider can become a CPRP resolver by implementing the following interface:

| Method | Input | Output | Purpose |
|--------|-------|--------|---------|
| `resolve` | namespace, name | `ResolutionResult` | Forward lookup: name → Party ID + metadata |
| `reverseResolve` | partyId | `ReverseResolutionResult` | Reverse lookup: Party ID → registered names |
| `resolveMulti` | queries[] | `ResolutionResult[]` | Batch resolution (latency optimization) |

> | `changelog` | since timestamp | `ChangelogEntry[]` | Credential changes since a given time |

Reviewer: Are these HTTP methods? Can you add references to their specification?

Each resolver publishes backing credentials via the CN Credentials Daml interface and reports its type:

| Resolver Type | Registration | Trust Tier | Example |
|--------------|-------------|-----------|---------|
| Featured | SV governance vote (on-ledger `ResolverFeaturedStatus`) | T3 | Freename, 7Trust |
| Permissionless | No registration required; implements the interface | T4 | Any third-party identity service |

> | Address Book | Local to the application; no network registration | App-defined | Citadel internal directory |

Reviewer: The `T3`, `T4` qualifiers introduce a bit of a cyclic reference. I wonder whether you can either write this CIP completely w/o such a forward reference on another CIP; or whether it is really one CIP being proposed, that just uses an extended format for its specification to make it more readable.

| Parallel | All resolvers queried simultaneously; results composed by weight | Maximum verification depth |
| Quorum | Resolution succeeds only when N resolvers agree on the same Party ID | High-security applications |

Two reference strategies are defined in the spec:
- `INSTITUTIONAL_DEFAULT` — parallel mode, strict collision policy, requires DNS or vLEI
- `PERMISSIVE_DEFAULT` — parallel mode, permissive collision policy, address-book-first

### Composition Engine

When multiple resolvers return results for the same query, the Composition Engine:

1. Groups results by `party_id`
2. For same-resolver duplicates: selects by Ledger Effective Time (last-write-wins, per CNS 2.0)
3. For cross-resolver duplicates: merges metadata, selects display name by weight
4. For conflicting `party_id` values: triggers collision handling per the strategy's `collision_policy`
5. Merges credential arrays, endpoint maps, and profile claims across all sources
6. Records per-claim provenance (`claim_sources`): for each metadata key, tracks which resolver and issuer contributed the value — enabling institutional audit trails

### Profile Rendering Guidelines

> Profile claims (`cns-2.0/name`, `cns-2.0/avatar`, `cns-2.0/email`, `cns-2.0/website`) are informational only and MUST NOT be interpreted as verified identity attributes. Verification status is determined exclusively by CIP-YYYY's trust evaluation, not by profile content. Display names SHOULD be ≤64 Unicode characters. Avatars SHOULD be `https://` URLs; applications MAY additionally support `ipfs://` URIs.

Reviewer: Unicode has a lot of typo-squatting space, and weird behavior in undefined ranges. Do you know what the latest guidance is for allowing Unicode in names?

Number: CIP-XXXX

Title: Canton Party Name Resolution Standard

Status: Draft

Author: Paolo Domenighetti (Freename), [Co-authors TBD]

Created: 2026-08-23

## Abstract

> This CIP defines the resolution layer for the Canton Network: how applications resolve human-readable names to Canton Party IDs and how they retrieve associated metadata. It specifies the Fully Qualified Party Name (FQPN) addressing format, the generic Resolver Interface that any identity provider implements, the resolution algorithm, the display conventions for rendered names, the built-in `party-id` resolver, guidelines for writing CIPs for additional resolvers, and a minimum fee model for on-ledger name state.

Reviewer: ```suggestion
This CIP defines the party name resolution layer for the Canton Network: how applications resolve human-readable names to Canton Party IDs and how they retrieve associated metadata. It specifies the Fully Qualified Party Name (FQPN) addressing format, the generic Resolver Interface that any identity provider implements, the resolution algorithm, the display conventions for rendered names, the built-in `party-id` resolver, guidelines for writing CIPs for additional resolvers, and a minimum fee model for on-ledger name state.
```

Reviewer: Question (may be answered later): why do we need a fee model?

> This CIP aligns with the CNS 2.0 architecture proposed by Simon Meier (Digital Asset) and the Working Group direction converged over the May–July 2026 meetings. Advanced name verification (a formal T1–T4 trust framework, per-source verification policies) is deferred to future work. Per-source import mechanics (DNS, LEI/vLEI, ENS, cross-chain) are specified in a companion CIP.

Reviewer: This feels out of place in the abstract. I'd push this into "Rationale".

## Motivation

> Canton participants are identified by cryptographic Party IDs — opaque strings of the form `<prefix>::<68-character-namespace-hash>`. These strings are unusable in any human workflow. CNS 1.0 names (`goldmansachs.unverified.cns`) are self-registered without verification, so they convey no signal about which counterparty they identify.

Reviewer: ```suggestion
Canton participants are identified by cryptographic Party IDs — opaque strings of the form `<prefix>::<68-character-public-key-hash>`. These strings are unusable in any human workflow other than copy-paste. CNS 1.0 names (`goldmansachs.unverified.cns`) are self-registered without verification, so they convey no signal about which counterparty they identify.
```

> The Identity and Metadata Working Group has identified three concrete gaps:

Reviewer: I'd remove this reference. It reads like a "proof by appeal to authority". I expect a CIP to justify its existence on its own.

- P1: Trustworthy human-readable names — applications need to display and accept names that map to specific Canton parties.
- P2: Off-ledger API endpoint discovery — applications need to find administrative endpoints (for example, the token-admin API for a CIP-56 instrument) keyed on a party identifier. A separate CIP-56 amendment (forthcoming from Digital Asset) will specify endpoint discovery on top of the CN Credentials Standard.
- P3: Self-published profile information — parties need to publish profile data that displays uniformly across applications, as specified in the Party Profile Credentials CIP (PixelPlex).

> This CIP addresses the resolution machinery underlying all three: a uniform addressing model, a pluggable resolver interface, and a shared display convention.

Reviewer: I don't see how P2 requires any part of the CPRP CIP. I also don't think P3 requires it. 

In my mind the motivation for the CPRP CIP is that there are many sources of names, and standardizing the mechanisms for resolving and rendering them avoids committing to a specific one and instead allows the ecosystem to adapt the sources that work best for the usecases at hand whenever the time is right.

## Specification

### 1. Terminology

> - Party ID. A Canton ledger identity of the form `<prefix>::<namespace>` where the namespace is a 68-character hexadecimal hash of the party administrator's public key. Party IDs may contain additional `::` separators introduced by administrator delegation.

Reviewer: Note that these are 68 characters is an implementation detail that may change. Can you loosen the qualification here?

> - Resolver. A software component that provides a mapping between names from an external or internal naming system and Canton Party IDs. Examples: `dns`, `lei`, `ens`, `cns`, `party-id`.

Reviewer: ```suggestion
- Resolver Name. The name of a system that provides a mapping between names from an external or internal naming system and Canton Party IDs. Examples: `dns`, `lei`, `ens`, `cns`, `party-id`.
```

> - Name Registry. On-ledger state, replicated in Canton via the CN Credentials Standard, that binds names within a resolver's scope to Party IDs.

Reviewer: Why does it have to be on-ledger for this CIP?

- Registrar. A party approved to operate a name registry (or a subset of it) for a given resolver. Registrars are governance concepts, not part of the FQPN addressing format.

> - FQPN. A Fully Qualified Party Name; see Section 2.

Reviewer: ```suggestion
- Fully Qualified Party Name (FQPN): ... add standalone definition here.
```

consider moving this key concept farther up

> - Rendered name. The human-readable form of an FQPN shown in application UIs; see Section 6.

Reviewer: ```suggestion
- Rendered name. The human-readable form of an FQPN shown in application UIs.
```

You don't have these sections for the other parts.

### 2. The Fully Qualified Party Name (FQPN)

#### 2.1 Format

A Fully Qualified Party Name is structured as three `:`-separated components:

```
<network>:<resolver>:<name>
```

| Component | Purpose | Example Values |
|-----------|---------|----------------|
| `network` | Prevents cross-environment confusion | `mainnet`, `testnet`, `devnet`, `localnet` |
| `resolver` | The naming system that backs this FQPN | `dns`, `lei`, `ens`, `cns`, `party-id` |
| `name` | A sequence of printable, non-whitespace ASCII characters, interpreted by the resolver | `blackrock.com`, `blackrock.canton`, `partyHint::hex-fingerprint` |

Examples:

- `mainnet:dns:blackrock.com`
- `mainnet:dns:tmmf.lloyds.com`
- `mainnet:cns:blackrock.canton`
- `mainnet:lei:784F5XWPLTWKTBV3E584`
- `mainnet:party-id:partyHint::1220abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000abcd0000`

> #### 2.2 Why free-form `<name>`

Reviewer: Move this into the "Rationale" section to keep the spec concise and unambiguous.

See https://github.com/canton-foundation/cips/blob/main/cip-0000/cip-0000.md#cip-format

The `<name>` component is a sequence of printable, non-whitespace ASCII characters. It is opaque to the resolution layer and interpreted only by the resolver referenced by the `<resolver>` component. This design allows any external naming syntax to be embedded without escaping: DNS names (`blackrock.com`), CNS names (`blackrock.canton`), LEI codes (`784F5XWPLTWKTBV3E584`), and Canton Party IDs with their `::` separators all fit unchanged in the `<name>` slot.

> The three-part format follows the direction proposed by Simon Meier in the June 25, 2026 CNS 2.0 next-steps presentation, after the June 18 Working Group discussion of whether a `registrar` component should appear in the FQPN. The direction: a registrar is a governance concept associated with a resolver, not a first-class addressing component. Adding a registrar slot would multiply icons in the UI, complicate the registrar lifecycle (what happens when a registrar changes for a namespace?), and offer little practical benefit to name discovery.

Reviewer: avoid appealing to authority. Instead provide the reason for why the decision was made this way.

#### 2.3 Network Discrimination

The `network` component is mandatory. TestNet names MUST NOT be confusable with MainNet names. Applications SHOULD display the network when they support multiple networks.

#### 2.4 The `party-id` Resolver (Built-in)

The `party-id` resolver is built into this CIP and available on every network. It returns an FQPN whose `<name>` component is the Canton Party ID verbatim. This resolver guarantees that every Canton party always has at least one FQPN, even before any name has been registered by any other resolver. The `party-id` FQPN has no external trust anchor and no display name beyond the Party ID prefix.

Example: `mainnet:party-id:partyHint::1220abcd...`

### 3. The Resolver Interface

Every resolver — built-in or third-party — implements the following logical interface. HTTP and gRPC mappings are conventional; see Section 8.

| Method | Inputs | Output | Purpose |
|--------|--------|--------|---------|

> | `resolve` | name | ResolutionResult | Forward lookup: name → Party ID + metadata |

Reviewer: what about fuzzy resolution for auto-complete?

| `reverseResolve` | party_id | ResolutionResult[] | Reverse lookup: Party ID → names known to this resolver |
| `resolveMulti` | name[] | ResolutionResult[] | Batched forward lookup |
| `changelog` | since-cursor | event stream | Subscribe to changes (additions, archivals, revocations) |

#### 3.1 ResolutionResult Schema

```
{
  "fqpn"          : "<network>:<resolver>:<name>",
  "party_id"      : "<canton-party-id>",
  "display_name"  : "<human-readable string>",
  "metadata"      : { /* claim keys and values */ },
  "valid_until"   : "<ISO-8601-timestamp>",
  "status"        : "OK" | "EXPIRED" | "NOT_FOUND"
}
```

The `display_name` is the resolver's canonical string for the name (typically the `<name>` component itself).

#### 3.2 Error Codes

| Code | Meaning |
|------|---------|
| 1000 | Name not found |
| 1001 | Resolver temporarily unavailable |
| 1002 | Malformed FQPN |
| 1003 | Network mismatch |

### 4. Name Resolution Algorithm

> An application resolving a user-typed name follows a small, deterministic algorithm. The default is intentionally simple; multi-source composition is an optional advanced feature (Section 5).

Reviewer: Can we avoid the optional feature? Optional features lead to ecosystem fragmentation into the parts that do support it and the ones that don't. Let's focus on the minimal parts that are required to work with FQPNs in their ASCII and graphical rendering.

Composition records per-claim provenance (`claim_sources`) tracking which resolver contributed each metadata value. When results agree on Party ID, metadata is merged with source attribution. When results disagree on Party ID, composition returns all candidates with their sources; composition does not itself pick a winner. The application decides how to present the disambiguation.

Composition is not the default resolution path. The simple resolution algorithm of Section 4 covers UI use cases, aligned with Leonid Rozenberg's observation (June 25, 2026) that name-selection UI in the wild is always context-scoped and never universally cross-source.

### 6. Display Conventions

Rendered names follow a shared convention: a source icon followed by a human-readable ASCII string.

```
<icon> <display-name>
```

The icon identifies the source of the name and is computed from the FQPN:

- For external-system resolvers (`dns`, `lei`, `ens`) and the built-in `party-id` resolver, there is one standardized icon per resolver.

> - For the `cns` resolver, the icon is the logo of the TLD under which the name is registered. Registrars self-publish their desired logo for their TLD, and the logo is registered via a DSO-issued credential (`cip-cns/tld-logo-url`, per the CNS 2.0 CIP). At render time, the application retrieves the logo URL from the DSO Credentials Registry and displays `<img src=logoUrl> <display-name>`.

Reviewer: Let's not define the CNS resolver in this standard.

#### 6.1 Asset Names

Assets are identified by the Canton token standard as `InstrumentId{admin: Party, id: Text}`. The rendered form for an asset combines the admin's rendered name with the asset symbol:

```
<asset-symbol> by <admin-rendered-name>
```

Example: `CC by <cns icon> dso.cns`, `BUIDL by <dns icon> blackrock.com`.

The asset symbol is obtained by:
1. Resolving the admin Party ID to its preferred FQPN using this CIP.
2. Retrieving the admin's off-ledger token-standard API endpoint via the CN Credentials Registry (per the forthcoming CIP-56 amendment).
3. Querying that API at the token-standard endpoint `/registry/metadata/v1/instruments/{id}` to obtain the symbol published by the admin.

Assets do not receive their own FQPNs; they are addressed by admin + id per the token standard, and rendered via this convention.

#### 6.2 Three Rendering Contexts

| Context | Where used | Content |
|---------|------------|---------|
| Inline | Transaction lists, counterparty fields | `<icon> <display-name>` |
| Profile summary | Hover tooltip, popover | Rendered name + resolver + published profile claims |

> | Full profile | Explorer profile page | Complete credential set, resolver history, changelog |

Reviewer: These two belong the the party profile CIP AFAIS.

> The full profile MAY be hosted by any explorer, not exclusively the Canton Foundation's Scan.

Reviewer: The profile will likely not be hosted on the SV node's Scan instance. Let's remove that hint.

### 7. Trust and Verification

This CIP intentionally does not define a formal trust framework. The initial trust model is minimal:

> - The DSO party issues credentials naming approved TLD registry admins and approved registrars per TLD (see the CNS 2.0 CIP led by Digital Asset).

Reviewer: TLD's are a CNS concept. They are not part of FQPNs AFAIS. So we shouldn't standardize these here.

- Each resolver's credibility follows from these DSO-issued credentials plus the resolver's own operational reputation.
- Applications trust resolvers by including or excluding them in their resolver configuration.

An advanced verification framework (four-tier T1–T4 model, per-source verification policies, application-configurable trust thresholds) is documented as future work in a companion draft (`cprp-verification.md`, retained as reference material). That framework is not part of the initial CIP set and is expected to be pursued once the resolution and import CIPs are stable.

### 8. Off-Ledger Resolution Service API and Availability

#### 8.1 API

The Resolver Interface (Section 3) is exposed by convention as an HTTPS + gRPC service.

| Endpoint | Method | Purpose |
|----------|--------|---------|
| `/v1/resolve` | POST | Single forward lookup |
| `/v1/resolve/batch` | POST | Batched forward lookup |
| `/v1/resolve/reverse` | POST | Reverse lookup |

> | `/v1/changelog` | GET (SSE) | Subscribe to changes |

Reviewer: A reference to an OpenAPI file for the full API would be great to have. It removes the ambiguity and creates value by precisely defining the API that we'd expect to mediate between apps and name resolvers.

Reviewer: You could create one as a PR against splice.

When a resolver is unreachable, an application's resolver configuration MAY fall back to the next resolver in its list; the application MUST NOT silently substitute a lower-trust resolver for a higher-trust one without explicit user or configuration consent. Applications SHOULD surface resolver-unavailable conditions to the user rather than presenting a stale-cache result as fresh. The `party-id` built-in resolver (Section 2.4) is always available as a last-resort fallback that returns the Party ID itself as its FQPN, since it requires no external state.

The off-ledger placement is a deliberate design choice to limit ACS growth on the DSO party and to keep the Scan API surface narrow. The tradeoff is that the resolution layer's availability depends on resolver operator infrastructure rather than being an inherent property of the ledger. Applications with settlement-critical dependencies on name resolution SHOULD provision or contract for resolver infrastructure at the same availability standard as their ledger connectivity.

### 9. On-Ledger Representation

> Name registrations are encoded as standard CN Credentials, using the credential registry defined by the CN Credentials Standard CIP (Digital Asset). This CIP does not introduce custom Daml templates.

Reviewer: I would explicitly not require this.

Reviewer: It limits the implementation freedom without clear gains.

What we do want to standardize on is how the names are represented (FQPNs), resolved, and rendered; and how additional name sources can be integrated into Canton Network by defining new resolvers or extending existing ones.

The `cprp/` claim-key prefix is a working prefix; it will be renamed to the `cip-<nr>/` form once this CIP receives its number, consistent with the claim-key convention used by the CNS 2.0 CIP (`cip-cns/*`) and the Party Profile Credentials CIP.

Import model. Names imported from external systems (DNS, LEI, ENS, and future importers) are materialized on-ledger as credentials so that Daml code can reference them. This follows the direction confirmed by Simon Meier in the WG meeting of July 9, 2026: "Import model is the only one that will allow records to be referenced by Daml code." Bridge-style references to external systems without on-ledger materialization are acceptable only for private or API-only resolvers where no Daml logic touches the name.

### 10. Fees

Name registration on-ledger consumes state and warrants a minimum fee model to prevent unbounded growth and to fund registrar operations. This CIP specifies the shape of the fee model, following the direction confirmed by Simon Meier at the WG meeting of July 9, 2026:

- Fees are paid per registered name.
- Read traffic (resolution queries) is free.
- Write traffic (registrations, updates, renewals) is paid via a fee.

> The recommended implementation reuses the extended-duration credential mechanism introduced by the CN Credentials Standard CIP (proposed default fee of $1/year, first 90 days free, paid by CC burn requiring authorization from the funds owner and the issuer or holder). Registrars MAY charge above the minimum for higher-value names; this pricing latitude is one mechanism for reducing name squatting. Full tokenomics is out of scope for this CIP and expected to be addressed in a future governance CIP.

Reviewer: Why do we need to standardize on these? The one thing that we should probably standardize on is to set the expectation that name resolution should by default be free within a fair use policy.

### 11. Guidelines for Writing CIPs for Additional Resolvers

Any project may propose a new resolver via its own CIP. Such a CIP MUST specify:

> - The resolver's `<resolver>` identifier as it appears in FQPNs.

Reviewer: ```suggestion
- The unique `<resolver name>` of the resolver as it appears in FQPNs.
```

- The `<name>` syntax accepted by this resolver.
- The naming authority and governance under which registrars operate.
- The verification procedure by which the resolver establishes name-to-Party-ID bindings.
- The credential encoding used to publish on-ledger name registrations.

> - The re-verification cadence, if the underlying authority state can change externally.

Reviewer: why do we need these?
